### ISSUE TITLE
Issue #25: persistent scratch buffers for measured hot-path allocations

### ISSUE THREAD

**Issue by Nando2392** [Nando2392/qx-mini-c#25]
# Issue #25 — persistent scratch buffers for measured hot-path allocations

## Baseline verified

- Branch: `main`
- HEAD/local and `origin/main`: `6a25a04f70b7a9161e4a6f4fda81eabef7e2d646`
- Open issues before creation: none
- Baseline CI: Issue #24 run `32476769189` success
- Tree was clean before causal instrumentation.

## Scope

Only CPU priority #2 from the roadmap: persistent scratch buffers to eliminate repeated `malloc`/`calloc`/`free` in the real inference hot path when profiling proves the seam is material.

Out of scope: threading, SIMD/AVX2/FMA, CUDA, expert cache, prefill GEMM, speculative decoding, KV2, QXF format changes, global pools/singletons, cross-process cache, or math changes.

## Roadmap reconciliation from Issue #24 evidence

Use `wiki/evidence/issue-24-qxf-mmap-baseline.json` as the source of truth, not stale roadmap numbers:

- F32 total buffered/mmap ratio: about `2.12548x`.
- `q8_k_compat` total buffered/mmap ratio: about `1.96456x`.
- mmap adds about `2.35 GB` peak RSS in that slice.
- Native prefill/decode did not show material improvement.
- Buffered remains default.
- Do not claim global improvement.

## Causal inventory before implementation

Static trace plus opt-in runtime allocation profiling identified material allocation churn in the current full-MoE/final-head path:

- `qx_acquire_span` / `qx_read_raw_span` (`src/qx_format.c:709-737` and wrappers): buffered backend allocates an owned span for every tensor slice read; mmap returns a borrowed pointer.
- `qx_rope_gqa_attention_partial` (`src/qx_format.c:4630-4758`): per layer/token allocates `qbuf`, `obuf`, `qfloat`, `context`, `scores`, `weights` and frees them before returning. Lifetime is one attention call. Persistent workspace can safely reuse these if no borrowed view survives reset.
- `qx_apply_real_moe_layer` (`src/qx_format.c:4761-4868`): per layer/token allocates a combined float buffer for `ffn_input`, `moe_output`, `expert_output`, `gate_values`, `up_values`, `expert_hidden`. Lifetime is one MoE layer call.
- `qx_packed_expert_matvec_mode` (`src/qx_format.c:4318-4365`): each expert projection calls `qx_read_raw_span`. In buffered mode this allocates and frees a row/expert slice; in mmap it does not allocate slice storage. This explains the large buffered vs mmap `malloc` delta.
- `qx_compute_real_final_head` (`src/qx_format.c:3216-3318`): F32 final head allocates/frees a `raw` chunk per 64 vocab rows via `qx_read_raw_span`; buffered allocates, mmap borrows.
- `qx_dump_prompt_state_loop_probe_summary` (`src/qx_format.c:5522-6152`): run-level buffers (`kbuf`, `vbuf`, `kcache`, `vcache`, `kfloat`, `vfloat`, scales, residual) already have execution lifetime; not priority scratch churn.
- `qx_apply_f32_rmsnorm`, `qx_apply_f32_head_rmsnorm`, embedding and router reads allocate via `qx_read_raw_span` mostly for tensor weights; persistent scratch can reduce buffered temporary allocation only if ownership is explicit and no stale pointer escapes.

## Opt-in profiling sample

Temporary causal instrumentation counted allocations inside `src/qx_format.c` for one warm command:

Command shape:

```text
build/qxqxf.exe prompt-state-loop-probe --in models/Qwen3-30B-A3B-UD-IQ2_M.qxf --tokenizer models/Qwen3-30B-A3B.qxt --text-file tests/fixtures/q8k_perf_prompt.txt --generate 1 --layers 48 --ctx 16 --kv int8 --activation <mode> --io-backend <backend> --temperature 0 --seed 7 --full-moe --final-head --bench
```

Observed counters:

| backend | activation | malloc | calloc | free | bytes_requested | elapsed_sec |
|---|---:|---:|---:|---:|---:|---:|
| buffered | f32 | 49138 | 197 | 54977 | 2638569912 | 17.083 |
| buffered | q8_k_compat | 49138 | 197 | 54977 | 2638569912 | 4.145 |
| mmap | f32 | 5744 | 197 | 54977 | 1600100792 | 17.047 |
| mmap | q8_k_compat | 5744 | 197 | 54977 | 1600100792 | 4.198 |

Interpretation:

- mmap removes many buffered span allocations, so the remaining scratch target differs by backend.
- Wall-clock is dominated by math for this single run; allocation reduction must be proven by counters and must not claim global speedup.
- A persistent scratch seam should first target the per-layer/token attention and MoE scratch buffers, and optionally a bounded reusable span buffer for buffered reads only if ownership/lifetime is fail-closed.

## Hypothesis to test

A per-execution scratch workspace, opt-in via `--scratch-policy ephemeral|persistent` with default `ephemeral`, can reduce repeated allocations/frees in attention/MoE/final-head hot path without changing output signatures. It may retain peak scratch capacity and increase RSS by the retained workspace size.

## TDD / acceptance gates

Create RED tests before product implementation for:

1. workspace init/destroy idempotence;
2. first reserve;
3. reuse without new allocation;
4. controlled growth;
5. overflow rejection;
6. allocation failure behavior;
7. logical reset and stale data handling;
8. cleanup after partial error;
9. pointer invalidation semantics;
10. ephemeral default preserved;
11. persistent opt-in observable in CLI/JSON;
12. exact outputs ephemeral vs persistent in F32;
13. exact outputs ephemeral vs persistent in `q8_k_compat`;
14. buffered and mmap compatibility;
15. allocation/free counters reduce in the targeted path;
16. benchmark JSON fails closed;
17. broad regression.

## Benchmark design

Produce `wiki/evidence/issue-25-persistent-scratch-baseline.json` with at least:

- 2 scratch policies: ephemeral, persistent
- 2 activations: f32, q8_k_compat
- primary backend: buffered
- secondary compatibility backend: mmap when feasible
- 1 warm-up and at least 3 measured repetitions per cell
- same executable, model, tokenizer, prompt, layers, ctx, KV, seed, token count and arguments
- startup/load, prefill, decode, total, tokens/s, peak RSS, allocations, frees, bytes requested, peak scratch capacity, growth events, raw runs, median, MAD, stdev, pstdev and complete provenance

Fail closed if any cell/provenance/signature/counter is missing, if persistent does not reduce targeted allocations, or if outputs drift within the same activation/backend.

## Documentation

Update:

- `wiki/comparisons/optimization-priorities.md`
- `wiki/queries/current-status-and-roadmap.md`
- new Issue #25 comparison page
- `wiki/index.md`
- new evidence JSON

Document allocations eliminated, allocations remaining, ownership/lifetime, RSS, limitations, whether the hypothesis was confirmed/refuted, and why priority 3 is not started yet.

**Comment by Nando2392** [comment]
COMPLETED in `ac53c5e4884e584284c06f7612c2d3590b814117`.

Evidence:
- Legal-diff-SHA256: `f1aee09fa3a457e8e7ff7f4991b23ee211a25cf40bd8addcf948b73be34779e9`
- CI: https://github.com/Nando2392/qx-mini-c/actions/runs/32485931790 — SUCCESS (`windows-msvc` 2m27s)
- New evidence JSON: `wiki/evidence/issue-25-persistent-scratch-baseline.json`
- Evidence matrix: 8 cells (`ephemeral|persistent` × `buffered|mmap` × `f32|q8_k_compat`), 1 warm-up + 3 measured repetitions per cell
- Persistent reduced median allocation churn in every cell: `480` mallocs and `672` frees fewer than ephemeral
- Persistent scratch peak: `65,536` bytes
- Output equality held exactly within each activation/backend: prompt token IDs, selected token, final token, residual/norm/logits checksums, K/V checksums, and `cache_readback_ok`
- Default remains `ephemeral`; `persistent` is opt-in via `--scratch-policy persistent`
- Buffered remains default I/O; mmap metrics from Issue #24 were reconciled in roadmap docs

Local verification:
- `cmd.exe /c build_msvc.bat` — PASS
- `python -m py_compile scripts/q8k_perf_experiment.py tests/test_q8k_perf_experiment.py smoke_check.py scripts/lint_wiki.py` — PASS
- `python smoke_check.py` — PASS before local WDAC re-block after rebuild
- `python scripts/lint_wiki.py` — PASS
- `tests/test_gguf.py tests/test_q8k_perf_experiment.py tests/test_qxf_mmap_api.py` — PASS (`12 passed`)
- WSL GCC C17 build — PASS
- WSL ASAN/UBSAN persistent probe — PASS (`final_token=358`, `scratch_peak_capacity_bytes=65536`)
- Full local Windows pytest was blocked by Smart App Control/WDAC (`WinError 4551`); CI Windows for the exact SHA passed.

Independent review:
- Security/lifecycle/ownership/overflow exact-digest review — PASS
- Benchmark/spec/provenance/claims exact-digest review — PASS
- Hermes Auto Research — `ready_for_review`, `stagedPayloadPresent=true`, digest matched, `modelReferencesVerified=true`, `canUseForProductChange=true`

Scope not claimed:
- No threading, SIMD, CUDA, expert cache, prefill GEMM, speculative decoding, or KV2.
- No global throughput claim and no promotion of persistent scratch to default.
